### DISCUSSION TITLE
Event log order

### DISCUSSION THREAD

**Discussion by arturodrt** [nasa/fprime#2041] [Q&A]
Is the execution order of fprime events reliable, or can it be affected by the fact that some components have independent threads?

Is it possible that the event log is affected by this?

**Comment by SterlingPeet** [comment]
The shortest answer to your question is: The concurrency order is defined by your hardware and OS architecture, and F Prime threads (active components) are subject to these limitations.

In an attempt to read between the lines, it sounds you might be trying to debug something involving resource contention on a system that is not real time, e.g.: Linux.  Depending on what you are trying to do, and your budget, and your personnel resources, there are two basic approaches to addressing resource contention issues:

1. Use an RTOS, and leverage its guarantees to address the resource contention
2. Design your topology and component to enforce execution order

The latter option can be accomplished in several ways, but I would tend to use an active rate group to component to indirectly control the access to the resource by making each contentious component process data on its own time, and then access the resource via a passive port call on the rate group input.  This works because the passive port call is executing in the calling component's execution context (thread on Linux) and doesn't return until the port handler execution is complete.  Therefore, only one component on the rate group can execute it's port handler at one time, and only while the active rate group component is calling it in sequence.  This has the added benefit of making guarantees about the order of execution as well.

My response here is extrapolating a lot, because your question is quite vague.  If you are able to provide us some more context, perhaps what OS and what types of resources or events you are debugging then we may be able to help you more effectively.
